cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -1256,6 +1318,24 @@ func generateMountsModeRun(mst *initramfsMountsState) error {
 		if !matches {
 			return fmt.Errorf("cannot validate boot: ubuntu-save mountpoint is expected to be from disk %s but is not", disk.Dev())
 		}
+
+		if isEncryptedDev {
+			// in run mode the path to open an encrypted save is for
+			// data to be encrypted and the save key in it
+			// to be successfully used. This already should stop
+			// allowing to chose ubuntu-data to try to access
+			// save. as safety boot also stops if the keys cannot
+			// be locked.
```

Comment:
this is fine, but I think we can make this a bit more clear:

```suggestion
			// in run mode, the path to open an encrypted ubuntu-save 
			// is for ubuntu-data to be decrypted and the save key in it
			// to be successfully used. This is already sufficient to stop
			// using an attacker-replaced ubuntu-data to try to access the
			// real ubuntu-save because to unlock ubuntu-save, the attacker
			// would also need to know the plaintext keys from decrypted 
			// ubuntu-data.
			// For additional safety, we stop boot if the keys cannot be locked.
```

I'm happy to take this as a follow-up in my pr which will clean up a number of related things here